python-sdk/nuimages/nuimages.py

```diff
@@ -0,0 +1,773 @@
+        """
+        Load an attribute and add it to self, if it isn't already loaded.
+        :param attr_name: The name of the attribute to be loaded.
+        :param loading_func: The function used to load it if necessary.
+        :return: The loaded attribute.
+        """
+        if attr_name in self.__dict__.keys():
+            return self.__getattribute__(attr_name)
+        else:
+            attr = loading_func(attr_name)
+            self.__setattr__(attr_name, attr)
+            return attr
+
+    def __load_table__(self, table_name) -> List[dict]:
+        """
+        Load a table and return it.
+        :param table_name: The name of the table to load.
+        :return: The table dictionary.
+        """
+        start_time = time.time()
+        table_path = osp.join(self.table_root, '{}.json'.format(table_name))
+        assert osp.exists(table_path), 'Error: Table %s does not exist!' % table_name
+        with open(table_path) as f:
+            table = json.load(f)
+        end_time = time.time()
+
+        # Print a message to stdout.
+        if self.verbose:
+            print("Loaded {} {}(s) in {:.3f}s,".format(len(table), table_name, end_time - start_time))
+
+        return table
+
+    def shortcut(self, src_table: str, tgt_table: str, src_token: str) -> Dict[str, Any]:
+        """
+        Convenience function to navigate between different tables that have one-to-one relations.
+        E.g. we can use this function to conveniently retrieve the sensor for a sample_data.
+        :param src_table: The name of the source table.
+        :param tgt_table: The name of the target table.
+        :param src_token: The source token.
+        :return: The entry of the destination table corresponding to the source token.
+        """
+        if src_table == 'sample_data' and tgt_table == 'sensor':
+            sample_data = self.get('sample_data', src_token)
+            calibrated_sensor = self.get('calibrated_sensor', sample_data['calibrated_sensor_token'])
+            sensor = self.get('sensor', calibrated_sensor['sensor_token'])
+
+            return sensor
+        elif (src_table == 'object_ann' or src_table == 'surface_ann') and tgt_table == 'sample':
+            src = self.get(src_table, src_token)
+            sample_data = self.get('sample_data', src['sample_data_token'])
+            sample = self.get('sample', sample_data['sample_token'])
+
+            return sample
+        else:
+            raise Exception('Error: Shortcut from %s to %s not implemented!' % (src_table, tgt_table))
+
+    def check_sweeps(self, filename: str) -> None:
+        """
+        Check that the sweeps folder was downloaded if required.
+        :param filename: The filename of the sample_data.
+        """
+        assert filename.startswith('samples') or filename.startswith('sweeps'), \
+            'Error: You passed an incorrect filename to check_sweeps(). Please use sample_data[''filename''].'
```

Comment:
One `'`in `sample_data[''filename'']` is sufficient?

Reply:
We need to escape the `'` character.